Smt/Reconstruct/Int/Polynorm.lean

```diff
@@ -60,8 +61,62 @@ def denote (ctx : Context) (m : Monomial) : Int :=
 theorem denote_neg {m : Monomial} : m.neg.denote ctx = -m.denote ctx := by
   simp only [neg, denote, Int.neg_mul_eq_neg_mul]
 
-theorem denote_mul {m₁ m₂ : Monomial} : (m₁.mul m₂).denote ctx = m₁.denote ctx * m₂.denote ctx :=
-  sorry
+theorem eq : ∀ {m n : Monomial}, m.coeff = n.coeff → m.vars = n.vars → m = n
+  | ⟨_, _⟩, ⟨_, _⟩, rfl, rfl => rfl
+
+
+
+
+section
+variable {op : α → α → α} (assoc : ∀ a b c, op (op a b) c = op a (op b c))
+
+-- Can be generalized to `List.foldl_assoc`.
+theorem foldl_assoc {g : β → α} (z1 z2 : α) :
+  List.foldl (fun z a => op z (g a)) (op z1 z2) l =
+  op z1 (List.foldl (fun z a => op z (g a)) z2 l) := by
+  revert z1 z2
+  induction l with
+  | nil => simp
+  | cons y ys ih =>
+    intro z1 z2
+    simp only [List.foldl_cons, ih, assoc]
+
+theorem foldr_assoc {g : β → α} (z1 z2 : α) :
+  List.foldr (fun z a => op a (g z)) (op z1 z2) l =
+  op z1 (List.foldr (fun z a => op a (g z)) z2 l) := by
+  revert z1 z2
+  induction l with
+  | nil => simp
+  | cons y ys ih =>
+    intro z1 z2
+    simp only [List.foldr_cons, ih, assoc]
+
+
+end
+-- Can be generazlized.
+theorem foldl_mul_insert {ctx : Context} :
+  List.foldl (fun z a => z * (ctx a)) 1 (mul.insert y ys) =
+  (ctx y) * List.foldl (fun z a => z * (ctx a)) 1 ys := by
+  induction ys with
+  | nil => simp [mul.insert]
+  | cons x ys ih =>
+    by_cases h : y ≤ x
+    · simp [mul.insert, h, foldl_assoc Int.mul_assoc (ctx y) (ctx x), Int.mul_assoc]
+    · simp only [mul.insert, h, List.foldl_cons, ite_false, Int.mul_comm,
+                 foldl_assoc Int.mul_assoc, ih]
+      rw [← Int.mul_assoc, Int.mul_comm (ctx x) (ctx y), Int.mul_assoc]
```

Comment:
I don't have a strong opinion (yet!) on whether or not a single whitespace should follow `←` inside `simp`. However, be consistent: either always add the whitespace or don't.